0000-spec-distribution.md

```diff
@@ -0,0 +1,307 @@
+
+Where:
+- O and P are buildpacks IDs referencing buildpacks that compose other buildpack IDs.
+- A through H are buildpack IDs referencing executable buildpacks. 
+
+Given:
+
+<img src="http://tex.s2cms.ru/svg/%0AO%20%3D%0A%5Cbegin%7Bbmatrix%7D%0AA%2C%20%26%20B%20%5C%5C%0AC%2C%20%26%20D%0A%5Cend%7Bbmatrix%7D%0A" alt="
+O =
+\begin{bmatrix}
+A, &amp; B \\
+C, &amp; D
+\end{bmatrix}
+" />
+
+<img src="http://tex.s2cms.ru/svg/%0AP%20%3D%0A%5Cbegin%7Bbmatrix%7D%0AE%2C%20%26%20F%20%5C%5C%0AG%2C%20%26%20H%0A%5Cend%7Bbmatrix%7D%0A" alt="
+P =
+\begin{bmatrix}
+E, &amp; F \\
+G, &amp; H
+\end{bmatrix}
+" />
+
+We propose:
+
+<img src="http://tex.s2cms.ru/svg/%0A%5Cbegin%7Bbmatrix%7D%0AE%2C%20%26%20O%2C%20%26%20F%0A%5Cend%7Bbmatrix%7D%20%3D%20%0A%5Cbegin%7Bbmatrix%7D%0AE%2C%20%26%20A%2C%20%26%20B%2C%20%26%20F%20%5C%5C%0AE%2C%20%26%20C%2C%20%26%20D%2C%20%26%20F%20%5C%5C%0A%5Cend%7Bbmatrix%7D%0A" alt="
+\begin{bmatrix}
+E, &amp; O, &amp; F
+\end{bmatrix} = 
+\begin{bmatrix}
+E, &amp; A, &amp; B, &amp; F \\
+E, &amp; C, &amp; D, &amp; F \\
+\end{bmatrix}
+" />
+
+<img src="http://tex.s2cms.ru/svg/%0A%5Cbegin%7Bbmatrix%7D%0AO%2C%20%26%20P%0A%5Cend%7Bbmatrix%7D%20%3D%20%0A%5Cbegin%7Bbmatrix%7D%0AA%2C%20%26%20B%2C%20%26%20E%2C%20%26%20F%20%5C%5C%0AA%2C%20%26%20B%2C%20%26%20G%2C%20%26%20H%20%5C%5C%0AC%2C%20%26%20D%2C%20%26%20E%2C%20%26%20F%20%5C%5C%0AC%2C%20%26%20D%2C%20%26%20G%2C%20%26%20H%20%5C%5C%0A%5Cend%7Bbmatrix%7D%0A" alt="
+\begin{bmatrix}
+O, &amp; P
+\end{bmatrix} = 
+\begin{bmatrix}
+A, &amp; B, &amp; E, &amp; F \\
+A, &amp; B, &amp; G, &amp; H \\
+C, &amp; D, &amp; E, &amp; F \\
+C, &amp; D, &amp; G, &amp; H \\
+\end{bmatrix}
+" />
+
+Note that buildpack IDs are expanded depth-first in left-to-right order.
+
+If a buildpack entry within a group has the parameter `optional = true`, the a copy of the group without the entry is repeated after the original group. This is functionally equivalent to the current meaning of optional for a single buildpack implementation. The current meaning of optional for a single buildpack implementation will remain unchanged, as it is more performant and results in the same behavior.
+
+
+## User Interface
+
+### App Developer
+
+`pack build` should accept a list of buildpacks via the `--buildpack` flag.
+The flag may be passed multiple times to construct a buildpack group.
+
+The value of each flag must be one of:
+- A buildpack ID of a buildpack on the builder, optionally followed by `@` and a version
+- A path to a buildpackage on the local filesystem
+- A reference to a buildpackage on a Docker registry  
+
+A version must be specified if the version is ambiguous.
+
+### Buildpack Developer
+
+`pack create-package` will package a selection of
+
+- gzip compressed, tar archived buildpack blobs
+- other buildpackages
+- stack and buildpack metadata
+
+into a `.cnb` file, OCI image in a registry, or OCI image in a Docker daemon.
+
+These properties will be specified in a `package.toml` file.
```

Comment:
This file looks very similar to `buildpack.toml`. Is it crazy to think maybe they should be the same?

Reply:
`buildpack.toml` defines buildpacks -- it has their metadata, names, paths to their source, etc.

`package.toml` references buildpacks. It's the (potentially ephemeral) configuration file for `pack create-package`. It contains real-world URIs, etc. that are used to build buildpackages. This is information that can't live in buildpack.toml.